apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ForwardIncludeRequestDispatcherInstrumentation.java

```diff
@@ -0,0 +1,129 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.servlet;
+
+import co.elastic.apm.agent.bci.ElasticApmInstrumentation;
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import co.elastic.apm.agent.impl.transaction.Span;
+import co.elastic.apm.agent.impl.transaction.TraceContextHolder;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.NamedElement;
+
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nullable;
+import javax.servlet.RequestDispatcher;
+import javax.servlet.ServletRequest;
+import javax.servlet.http.HttpServletRequest;
+import java.util.Arrays;
+import java.util.Collection;
+
+import static net.bytebuddy.matcher.ElementMatchers.*;
+
+public class ForwardIncludeRequestDispatcherInstrumentation extends ElasticApmInstrumentation {
+
+    private static final String SPAN_TYPE = "servlet";
+    private static final String SPAN_SUBTYPE = "request-dispatcher";
+    private static final String FORWARD_SPAN_ACTION = "forward";
+    private static final String INCLUDE_SPAN_ACTION = "include";
+    private static final String FORWARD = "FORWARD";
+    private static final String INCLUDE = "INCLUDE";
+
+    @Override
+    public Class<?> getAdviceClass() {
+        return ForwardIncludeRequestDispatcherAdvice.class;
+    }
+
+    public static class ForwardIncludeRequestDispatcherAdvice extends ForwardIncludeRequestDispatcherInstrumentation {
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        private static void beforeExecute(@Advice.Local("span") @Nullable Span span,
+                                          @Advice.Origin("#m") String method,
+                                          @Advice.Argument(0) @Nullable ServletRequest request) {
+            if (tracer == null || tracer.getActive() == null) {
+                return;
+            }
+            final TraceContextHolder<?> parent = tracer.getActive();
+
+            if (request != null && request instanceof HttpServletRequest) {
+                HttpServletRequest httpServletRequest = (HttpServletRequest) request;
+                String servletPath = "";
+                String pathInfo = "";
+
+                if (FORWARD.equalsIgnoreCase(method)) {
+                    servletPath = (String) httpServletRequest.getAttribute(RequestDispatcher.FORWARD_SERVLET_PATH);
```

Comment:
Didn't you say these properties are not set yet? Or are they only set on the real application servers but were not on the unit tests?

Reply:
I tested  with Apache Tomcat Version 8.5.5, and those properties were not there.
They were not on the unit tests, too.
I don't know, if I should  instrument RequestDispatcher .

Reply:
I think there was a confusion here- the attribute `RequestDispatcher.FORWARD_SERVLET_PATH` provides the servlet path of the original request, but here we are interested in the new request paths, so just `HttpServletRequest#getServletPath` and `HttpServletRequest#getPathInfo`.